exec: a distributed query hangs when runtime error is encountered

Found when running logic tests with `vectorize=experimental_on`.

On the following logic test:
```
# LogicTest: 5node-dist-vec

statement ok
CREATE TABLE c (c_id INT PRIMARY KEY, bill TEXT);
CREATE TABLE o (o_id INT PRIMARY KEY, c_id INT, ship TEXT);
INSERT INTO c VALUES
    (1, 'CA'),
    (2, 'TX'),
    (3, 'MA');
INSERT INTO o VALUES
    (10, 1, 'CA'),
    (20, 2, 'TX'),
    (30, 2, NULL);
ALTER TABLE c SPLIT AT VALUES (2), (3);
ALTER TABLE o SPLIT AT VALUES (10), (20), (30);
ALTER TABLE c EXPERIMENTAL_RELOCATE VALUES
    (ARRAY[2], 1), (ARRAY[3], 2), (ARRAY[3], 3);
ALTER TABLE o EXPERIMENTAL_RELOCATE VALUES
    (ARRAY[1], 10), (ARRAY[2], 20), (ARRAY[3], 30);

query TTTI colnames
SELECT start_key, end_key, replicas, lease_holder from [SHOW EXPERIMENTAL_RANGES FROM TABLE c]
----
start_key  end_key  replicas  lease_holder
NULL       /2       {2}       2
/2         /3       {3}       3
/3         NULL     {3}       3

query TTTI colnames
SELECT start_key, end_key, replicas, lease_holder from [SHOW EXPERIMENTAL_RANGES FROM TABLE o]
----
start_key  end_key  replicas  lease_holder
/10        /20      {1}       1
/20        /30      {2}       2
NULL       /10      {3}       3
/30        NULL     {3}       3

query T
SELECT url FROM [EXPLAIN (DISTSQL) SELECT * FROM c WHERE EXISTS(SELECT * FROM o WHERE o.c_id=c.c_id)];
----
https://cockroachdb.github.io/distsqlplan/decode.html#eJzElEFvmzAUx-_7FOidWskIbJI2Q5rEpdNSbcnU5jZxcPFbQEoxso20qMp3nwKTEhizkxGlR2J-z--93z-8QSkFLvgraoh_AAUCDAhEkBKolMxQa6n2R-2Lc_EL4pBAUVa12f-cEsikQojfwBRmgxDDir9s8Am5QBVEQECg4cWmKV-p4pWrbSKBwHPFSx17AQ39gIWQ7gjI2vwpe6j2svVyrvNunYRBuksJaMPXCDHbkX-0d6hTl1IJVCg6ldI96XplYMYvXOePsihRBffd1lbbCmPv68Pnlff88G3uPS7nCyCwwZ_mJqG3n1Sxzs1Nwm6BwLI2sZdQ0oxzPP5htGjEaAN9L6Qvq2DW38Hg1ZPO1fR06eGw9Owg3Q_6E7uE045w-r-tTdx5ZKEfRJfMI71uHqdXzCM7ffPUGYqA-RcMxRmtMWco_ICGHi-FRz1pclQXTAe7bjru3ulrNdDVE-pKlhpP-hiF-7lQrLHdk5a1yvC7kllzTfu4bLjmDydQm_Z02j7My-ao2f3p8GwMTO0w7cPhMcw6MO3DzAp_7MBhH47OWNhfN9vh2RiY2uHJGM922OHZDjs8T8d4vhujyg47VNlhh6r7MarssEOVHXaomp2lKt19-B0AAP__OLixzQ==

query IT rowsort
SELECT * FROM c WHERE EXISTS(SELECT * FROM o WHERE o.c_id=c.c_id);
----
1  CA
2  TX
```
The last query encounters an error that is fixed by #39294, but it appears to be hanging when cleaning up the established infrastructure (i.e. the error is returned, but the flow shutdown doesn't succeed):
```
SELECT * FROM c WHERE EXISTS(SELECT * FROM o WHERE o.c_id=c.c_id);;
	-- FAIL
--- done: testdata/logic_test/temp with config 5node-dist-vec: 5 tests, 1 failures
^Cmake: *** [testbaselogic] Interrupt: 2
```
I thought maybe #39248 would fix the issue, but it doesn't seem to be the case.

Here is the plan:
<img width="701" alt="Screen Shot 2019-08-03 at 12 26 32 PM" src="https://user-images.githubusercontent.com/5135385/62416125-01484900-b5ea-11e9-824c-0995905dc599.png">

Interestingly, if we relocate all data of table `c` onto a single node (using this command
```
ALTER TABLE c EXPERIMENTAL_RELOCATE VALUES
    (ARRAY[3], 1), (ARRAY[3], 2), (ARRAY[3], 3);
```
), the hangup doesn't happen.
The plan is then 
<img width="713" alt="Screen Shot 2019-08-03 at 12 26 39 PM" src="https://user-images.githubusercontent.com/5135385/62416148-479da800-b5ea-11e9-953e-ff6884f06121.png">

--- a/pkg/sql/distsqlrun/column_exec_setup.go
+++ b/pkg/sql/distsqlrun/column_exec_setup.go
@@ -1080,7 +1080,7 @@ func (s *vectorizedFlowCreator) setupRemoteOutputStream(
 	atomic.AddInt32(&s.numOutboxes, 1)
 	run := func(ctx context.Context, cancelFn context.CancelFunc) {
 		outbox.Run(ctx, s.nodeDialer, stream.TargetNodeID, s.flowID, stream.StreamID, cancelFn)
-		atomic.AddInt32(&s.numOutboxes, -1)
+		currentOutboxes := atomic.AddInt32(&s.numOutboxes, -1)
 		// When the last Outbox on this node exits, we want to make sure that
 		// everything is shutdown; namely, we need to call cancelFn if:
 		// - it is the last Outbox
@@ -1089,7 +1089,7 @@ func (s *vectorizedFlowCreator) setupRemoteOutputStream(
 		// - cancelFn is non-nil (it can be nil in tests).
 		// Calling cancelFn will cancel the context that all infrastructure on this
 		// node is listening on, so it will shut everything down.
-		if atomic.LoadInt32(&s.numOutboxes) == 0 && !s.materializerAdded && cancelFn != nil {
+		if currentOutboxes == 0 && !s.materializerAdded && cancelFn != nil {
 			cancelFn()
 		}
 	}
@@ -1337,7 +1337,7 @@ func (s *vectorizedFlowCreator) setupOutput(
 			s.syncFlowConsumer,
 			metadataSourcesQueue,
 			outputStatsToTrace,
-			s.getCancelFlowFn(),
+			s.getCancelFlowFn,
 		)
 		if err != nil {
 			return err

--- a/pkg/sql/distsqlrun/flow.go
+++ b/pkg/sql/distsqlrun/flow.go
@@ -456,9 +456,6 @@ func (f *Flow) setupProcessors(ctx context.Context, inputSyncs [][]RowSource) er
 
 func (f *Flow) setup(ctx context.Context, spec *distsqlpb.FlowSpec) error {
 	f.spec = spec
-	ctx, f.ctxCancel = contextutil.WithCancel(ctx)
-	f.ctxDone = ctx.Done()
-
 	if f.isVectorized {
 		log.VEventf(ctx, 1, "setting up vectorize flow %d", f.id)
 		acc := f.EvalCtx.Mon.MakeBoundAccount()
@@ -490,6 +487,9 @@ func (f *Flow) startInternal(ctx context.Context, doneFn func()) error {
 	log.VEventf(
 		ctx, 1, "starting (%d processors, %d startables)", len(f.processors), len(f.startables),
 	)
+
+	ctx, f.ctxCancel = contextutil.WithCancel(ctx)
+	f.ctxDone = ctx.Done()
 
 	// Only register the flow if there will be inbound stream connections that
 	// need to look up this flow in the flow registry.

--- a/pkg/sql/distsqlrun/materializer.go
+++ b/pkg/sql/distsqlrun/materializer.go
@@ -56,9 +56,12 @@ type materializer struct {
 	// pass it down further). This allows for the cancellation of the tree rooted
 	// at this materializer when it is closed.
 	ctxCancel context.CancelFunc
-	// cancelFlow will cancel the context of the flow (i.e. if non-nil, it is
-	// Flow.ctxCancel).
-	cancelFlow context.CancelFunc
+	// cancelFlow will return a function to cancel the context of the flow. It is
+	// a function in order to be lazily evaluated, since the context cancellation
+	// function is only available when Starting. This function differs from
+	// ctxCancel in that it will cancel all components of the materializer's flow,
+	// including those started asynchronously.
+	cancelFlow func() context.CancelFunc
 }
 
 const materializerProcName = "materializer"
@@ -70,9 +73,10 @@ const materializerProcName = "materializer"
 // - metadataSourcesQueue are all of the metadata sources that are planned on
 // the same node as the materializer and that need to be drained.
 // - outputStatsToTrace (when tracing is enabled) finishes the stats.
-// - cancelFlow is the context cancellation function that cancels the context
-// of the flow (i.e. it is Flow.ctxCancel). It should only be non-nil in case
-// of a root materializer (i.e. not when we're wrapping a row source).
+// - cancelFlow should return the context cancellation function that cancels
+// the context of the flow (i.e. it is Flow.ctxCancel). It should only be
+// non-nil in case of a root materializer (i.e. not when we're wrapping a row
+// source).
 func newMaterializer(
 	flowCtx *FlowCtx,
 	processorID int32,
@@ -85,7 +89,7 @@ func newMaterializer(
 	output RowReceiver,
 	metadataSourcesQueue []distsqlpb.MetadataSource,
 	outputStatsToTrace func(),
-	cancelFlow context.CancelFunc,
+	cancelFlow func() context.CancelFunc,
 ) (*materializer, error) {
 	m := &materializer{
 		input:               input,
@@ -195,7 +199,7 @@ func (m *materializer) InternalClose() bool {
 	if m.ProcessorBase.InternalClose() {
 		m.ctxCancel()
 		if m.cancelFlow != nil {
-			m.cancelFlow()
+			m.cancelFlow()()
 		}
 		return true
 	}

--- a/pkg/sql/distsqlrun/vectorized_flow_shutdown_test.go
+++ b/pkg/sql/distsqlrun/vectorized_flow_shutdown_test.go
@@ -280,7 +280,7 @@ func TestVectorizedFlowShutdown(t *testing.T) {
 					nil, /* output */
 					materializerMetadataSources,
 					nil, /* outputStatsToTrace */
-					cancelLocal,
+					func() context.CancelFunc { return cancelLocal },
 				)
 				require.NoError(t, err)
 				materializer.Start(ctxLocal)

